frontend/components/tree/js/tree.service.ts

```diff
@@ -1015,43 +1022,121 @@ export class TreeService {
 
-		if (this.currentSelectedNodes) {
-			for (let i = 0; i < this.currentSelectedNodes.length; i++) {
-				this.currentSelectedNodes[i].selected = false;
+		if (this.currentSelectedNodes.length) {
+			while (this.currentSelectedNodes.length) {
+				const currentNode = this.currentSelectedNodes.pop();
+				currentNode.selected = this.SELECTION_STATES.unselected;
+				this.setSelectionOnParentNodes(currentNode, this.SELECTION_STATES.unselected);
+				if (currentNode.children && currentNode.children.length) {
+					this.currentSelectedNodes = this.currentSelectedNodes.concat(currentNode.children);
+				}
 			}
 		}
 
 		this.currentSelectedNodes = [];
 	}
 
+	/**
+	 * Set the selection state of a nodes parents to represent it's children
+	 * @param nodes the nodes to set a new selection state on
+	 * @param select wether the node should be selected or not
+	 */
+	public traverseNodesAndSetSelected(nodes: any[], select: boolean) {
+
+		nodes = nodes.concat(); // make a copy
+		const setParentNodes = nodes.concat();
+		let currentNode;
+
+		while (nodes.length) {
+			currentNode = nodes.pop();
+
+			const nodeIndex = this.currentSelectedNodes.indexOf(currentNode);
+			const notParentOfUnselected = currentNode.selected !== this.SELECTION_STATES.parentOfUnselected;
+
+			if (select && notParentOfUnselected && nodeIndex === -1) {
+				this.currentSelectedNodes.push(currentNode);
+			} else if (nodeIndex > -1) {
+				this.currentSelectedNodes.splice(nodeIndex, 1);
+			}
+
+			if (currentNode.toggleState !== "invisible") {
+
+				if (currentNode.toggleState !== "parentOfInvisible" && select) {
+					currentNode.selected = this.SELECTION_STATES.selected;
+				} else if (!select) {
+					currentNode.selected = this.SELECTION_STATES.unselected;
+				} else {
+					currentNode.selected = this.SELECTION_STATES.parentOfUnselected;
+				}
+
+				if (currentNode.children && currentNode.children.length) {
+					nodes = nodes.concat(currentNode.children);
+				}
+			}
+
+		}
+
+		setParentNodes.forEach((node) => {
+			this.setSelectionOnParentNodes(node);
+		});
+
+	}
+
+	/**
+	 * Set the selection state of a nodes parents to represent it's children
+	 * @param currentNode the node who's parents to traverse up
+	 * @param forceState a selected state to enforce upon parents
+	 */
+	public setSelectionOnParentNodes(currentNode: any, forceState?: string) {
+		const parentPath = this.getPath(currentNode._id);
+		parentPath.pop(); // Remove the node itself
+
+		for (let i = parentPath.length - 1; i >= 0; i--) {
+			const parentNode = this.getNodeById(parentPath[i]);
+
+			if (forceState) {
+				parentNode.selected = forceState;
+				continue;
+			}
+
+			let allUnselected = true;
+			let allSelected = true;
+
+			parentNode.children.forEach((n) => {
+				allUnselected = allUnselected && (n.selected === this.SELECTION_STATES.unselected || n.selected === undefined);
+				allSelected = allSelected && n.selected === this.SELECTION_STATES.selected;
```

Comment:
If you rewrite this into a standard forloop, you can potentially reduce the # iterations by breaking once `!allUnselected && !allSelected`

I.e. you can stop as soon as you realised it's parentOfUnselected

Whether it's worth it or not i'm not sure.